test.py

```diff
@@ -0,0 +1,311 @@
+
+        with self.assertRaisesRegex(ValueError, "safe_eval doesn't permit you to read gather_secret"):
+            exec(expr_checker("a.gather_secret()", ast_get_attr, check_type=check_type))
+
+    # FIXME: Need to implement ast_set_attr
+    # def test_readonly_type(self):
+    #     a = ReadOnlyObject(42)
+    #
+    #     exec(expr_checker("a.set_x_value(25)", ast_get_attr, check_type=check_type))
+    #
+    #     self.assertEqual(a.x, 42)
+
+    def test_delete_attr(self):
+        with self.assertRaisesRegex(ValueError, "safe_eval: doesn't permit you to delete attributes"):
+            exec(expr_checker("del a.x", ast_get_attr, check_type=check_type))
+
+    def test_method_return(self):
+        exec(expr_checker("Good().tell_me_hi()", ast_get_attr, check_type=check_type)),
+
+        with self.assertRaisesRegex(ValueError, "safe_eval doesn't permit you to read gift_of_satan"):
+            exec(expr_checker("Good().gift_of_satan()", ast_get_attr, check_type=check_type))
+
+    def test_object_with(self):
+        code = cleandoc(
+            """
+            with Good() as g:
+                g.tell_me_hi()
+            """
+        )
+
+        exec(expr_checker(code, ast_get_attr, check_type=check_type))
+
+        with self.assertRaisesRegex(ValueError, "safe_eval doesn't permit you to read gift_of_satan"):
+            code = cleandoc(
+                """
+                with Good() as g:
+                    g.gift_of_satan()
+                """
+            )
+            exec(expr_checker(code, ast_get_attr, check_type=check_type))
+
+        with self.assertRaisesRegex(ValueError, "safe_eval didn't like <__main__.Dangerous object at .+>"):
+            code = cleandoc(
+                """
+                with Dangerous() as d:
+                    pass
+                """
+            )
+            exec(expr_checker(code, ast_get_attr, check_type=check_type))
+
+    def test_function_return(self):
+        def foo():
+            return Good()
+
+        def anti_foo():
+            return Dangerous()
+
+        exec(expr_checker("foo()", ast_get_attr, check_type=check_type))
+
+        with self.assertRaisesRegex(ValueError, "safe_eval didn't like <__main__.Dangerous object at .+>"):
+            exec(expr_checker("anti_foo()", ast_get_attr, check_type=check_type))
+
+    def test_evil_decorator(self):
+        def evil(func):
+            return lambda: Dangerous()
+
+        @evil
+        def foo():
+            pass
+
+        with self.assertRaisesRegex(ValueError, "safe_eval didn't like <__main__.Dangerous object at .+>"):
+            code = expr_checker("foo()", ast_get_attr, check_type=check_type)
+            exec(code)
+            print(code)
+
+    def test_evil_exec(self):
+        with self.assertRaisesRegex(ValueError, "safe_eval didn't like 3.1415"):
+            exec(expr_checker("eval('3.1415')", ast_get_attr, check_type=check_type))
```

Comment:
looks like this raise the wrong error...

`'3.1415'` is a valid type isn't it ?

Reply:
It's a float and it's not on the whitelist :sweat_smile:

Reply:
so this test is not useful I think (you already test similar cases)